File: verifiers/python/test_pseudonym_refs.py
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).with_name("src")))

from allowly_receipt_format import matches_ref


KEY = bytes(range(32))
RECORD_REF = (
    "hmac-v1:ecfc67ffb7bac447c05df24d1a25d75e"
    "be7e765320d0fb1b4d22be332341599e"
)
FULL_TUPLE_REF = (
    "hmac-v1:c5fa890e042ea330013bd07bcc47c331"
    "2136373493fe87a490907b0e93c5727a"
)


def test_matches_published_record_vector() -> None:
    assert matches_ref(KEY, "record", "MRN-48291", RECORD_REF)


def test_matches_published_full_tuple_vector() -> None:
    value = "\x1f".join(
        ("MRN-48291", "baseline_arm_1", "demographics", "demographics", "2")
    )
    assert matches_ref(KEY, "full_tuple", value, FULL_TUPLE_REF)


def test_mismatch_or_noncanonical_reference_returns_false() -> None:
    cases = [
        ("record", "MRN-48292", RECORD_REF),
        ("actor", "MRN-48291", RECORD_REF),
        ("record", "MRN-48291", RECORD_REF.upper()),
        ("record", "MRN-48291", RECORD_REF[:-1]),
        ("record", "MRN-48291", "sha256:" + RECORD_REF.removeprefix("hmac-v1:")),
    ]
    for field_name, value, ref in cases:
        assert not matches_ref(KEY, field_name, value, ref)


def test_value_is_not_unicode_normalized() -> None:
    composed = "Jos\u00e9"
    decomposed = "Jose\u0301"
    ref = "hmac-v1:76eca5c20d6d5d1c60f48bcd8b891f757e21a030e92f3cac90bc2462997f43ab"
    assert matches_ref(KEY, "actor", composed, ref)
    assert not matches_ref(KEY, "actor", decomposed, ref)


def test_rejects_weak_key_or_unknown_field() -> None:
    try:
        matches_ref(b"short", "record", "MRN-48291", RECORD_REF)
    except ValueError as exc:
        assert "128 bits" in str(exc)
    else:
        raise AssertionError("weak pseudonym key was accepted")

    try:
        matches_ref(KEY, "unknown", "MRN-48291", RECORD_REF)
    except ValueError as exc:
        assert "field name" in str(exc)
    else:
        raise AssertionError("unknown pseudonym field was accepted")


def main() -> None:
    test_matches_published_record_vector()
    test_matches_published_full_tuple_vector()
    test_mismatch_or_noncanonical_reference_returns_false()
    test_value_is_not_unicode_normalized()
    test_rejects_weak_key_or_unknown_field()
    print("hmac-v1 pseudonym reference vectors pass.")


if __name__ == "__main__":
    main()

File: verifiers/typescript/test_pseudonym_refs.ts
/**
 * hmac-v1 pseudonym reference tests (spec Appendix A).
 *
 * Constants are the spec's Appendix A.3 published vectors — identical to the
 * Python test_pseudonym_refs.py, so the two reference helpers are pinned to the
 * same cross-language expected bytes.
 *
 * Usage: node --experimental-strip-types test_pseudonym_refs.ts
 *    or after build: node dist/test_pseudonym_refs.js
 */
import { matchesRef } from "./verifier.js";

const KEY = new Uint8Array(Array.from({ length: 32 }, (_, i) => i));
const RECORD_REF =
  "hmac-v1:ecfc67ffb7bac447c05df24d1a25d75ebe7e765320d0fb1b4d22be332341599e";
const FULL_TUPLE_REF =
  "hmac-v1:c5fa890e042ea330013bd07bcc47c3312136373493fe87a490907b0e93c5727a";

let failures = 0;
function check(name: string, cond: boolean): void {
  if (cond) {
    console.log(`  OK    ${name}`);
  } else {
    console.log(`  FAIL  ${name}`);
    failures++;
  }
}

// A.3 published record vector
check("matches_published_record_vector", matchesRef(KEY, "record", "MRN-48291", RECORD_REF));

// A.3 published full_tuple vector (five components joined by 0x1F)
const fullTuple = ["MRN-48291", "baseline_arm_1", "demographics", "demographics", "2"].join("\x1f");
check("matches_published_full_tuple_vector", matchesRef(KEY, "full_tuple", fullTuple, FULL_TUPLE_REF));

// Mismatches and non-canonical references return false (never throw)
for (const [fieldName, value, ref] of [
  ["record", "MRN-48292", RECORD_REF],
  ["actor", "MRN-48291", RECORD_REF],
  ["record", "MRN-48291", RECORD_REF.toUpperCase()],
  ["record", "MRN-48291", RECORD_REF.slice(0, -1)],
  ["record", "MRN-48291", "sha256:" + RECORD_REF.slice("hmac-v1:".length)],
] as const) {
  check(`mismatch_returns_false [${fieldName}/${value.slice(0, 9)}]`, !matchesRef(KEY, fieldName, value, ref));
}

// Value is NOT Unicode-normalized: composed U+00E9 matches, decomposed e+U+0301 does not.
const composed = "José";
const decomposed = "José";
const JOSE_REF = "hmac-v1:76eca5c20d6d5d1c60f48bcd8b891f757e21a030e92f3cac90bc2462997f43ab";
check("composed_value_matches", matchesRef(KEY, "actor", composed, JOSE_REF));
check("decomposed_value_does_not_match", !matchesRef(KEY, "actor", decomposed, JOSE_REF));

// Weak key and unknown field are rejected loudly
function throws(fn: () => void, needle: string): boolean {
  try {
    fn();
    return false;
  } catch (e) {
    return e instanceof Error && e.message.includes(needle);
  }
}
check("rejects_weak_key", throws(() => matchesRef(new Uint8Array(8), "record", "MRN-48291", RECORD_REF), "128 bits"));
check("rejects_unknown_field", throws(() => matchesRef(KEY, "unknown", "MRN-48291", RECORD_REF), "field name"));

if (failures) {
  console.log(`\n${failures} failure(s)`);
  process.exit(1);
}
console.log("\nhmac-v1 pseudonym reference vectors pass.");
